### PR TITLE
Add publishing

### PR DESCRIPTION
**Pull request by LuxMiranda**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -29,3 +29,8 @@ runTests
 # Vim 
 ##
 .swp
+.swo
```
Comment (nick-benoit14) [inline .gitignore]: `.swp` needs to be `*.swp`. That way you can remove things like `.cpp.swp`, which I also am not sure even works at all. This was fixed on the last PR though, so these changes should be removed.

File: lib/component.cpp [hunk 2]
```diff
@@ -1,10 +1,118 @@
...
 {
-  if (message == nullptr || communicator == nullptr)
+  if (message == nullptr)
   {
+    std::cout << "Invalid subscription request" << std::endl;
+  }
+
+  auto castMessage = std::dynamic_pointer_cast<SpaSubscriptionRequest>(message);
+
+  // Dynamic memory allocation !!!
```
Comment (nick-benoit14) [inline lib/component.cpp]: This comment is probably not necessary, we are allocating memory in lots of places.

File: .gitignore [hunk 3]
```diff
@@ -28,5 +28,13 @@ runTests
 ##
 # Vim
 ##
+.swp
```
Comment (nick-benoit14) [inline .gitignore]: Only `*.swp` and `*.swo` are necessary. The others dont actually do anything

File: lib/component.cpp [hunk 4]
```diff
@@ -72,16 +79,41 @@ void Component::receiveMessage(std::shared_ptr<SpaMessage> message)
   {
     return;
   }
-  else if (message->spaHeader.opcode == op_SPA_SUBSCRIPTION_REQUEST)
+
+  switch (message->spaHeader.opcode)
```
Comment (nick-benoit14) [inline lib/component.cpp:83]: I was going to suggest making this a switch statement. I like that better.

File: lib/component.cpp [hunk 5]
```diff
@@ -72,16 +79,41 @@ void Component::receiveMessage(std::shared_ptr<SpaMessage> message)
...
+    default:
+      std::cout << "Unrecognized message" << std::endl;
+      break;
   }
-  else if (message->spaHeader.opcode == op_SPA_SUBSCRIPTION_REPLY)
+}
+
+void Component::publish()
+{
+  for (auto i = 0u; i < subscribers.size(); ++i)
```
Comment (nick-benoit14) [inline lib/component.cpp:105]: Is `0u` a typo?

Comment (LuxMiranda) [reply lib/component.cpp:105]: Nope! It makes `i` an unsigned integer so it doesn't throw a warning when you compare it with the vector size.

File: lib/component.hpp [hunk 6]
```diff
@@ -43,8 +65,9 @@ class Component
 protected:
   LogicalAddress address;
   Com communicator;
-
-private:
+  uint8_t publishIter;
   uint16_t dialogId;
+  std::vector<Subscriber> subscribers; // Should we make this a vector of pointers?
```
Comment (nick-benoit14) [inline lib/component.hpp:70]: I think so, we can do that later on though

File: lib/component.cpp (hunk 5 above)
Comment (nick-benoit14) [reply lib/component.cpp:105]: Oh nice!

**Review by nick-benoit14 — APPROVED** [review]
:D

**Pull request merged**